File: QA-Assignment/BusinessWorkflowAutomation/pages/login_page.py
import re
import time
from playwright.sync_api import Page, expect
from pages.base_page import BasePage


class LoginPage(BasePage):
    """Page Object Model for the Login Page (/login)."""

    def __init__(self, page: Page):
        super().__init__(page)
        self.heading = page.get_by_role("heading", name="Account Login")
        self.email_input = page.get_by_test_id("login-email")
        self.password_input = page.get_by_test_id("login-password")
        self.login_button = page.get_by_test_id("login-button")
        self.signup_link = page.get_by_test_id("signup-button")

    def goto(self):
        """Navigate to Login Page and verify heading is displayed."""
        self.navigate_to("/login")
        expect(self.heading).to_be_visible(timeout=15000)

    def login(self, email: str = "", password: str = ""):
        """Perform login with given credentials."""
        self.email_input.clear()
        if email:
            self.email_input.fill(email)

        self.password_input.clear()
        if password:
            self.password_input.fill(password)

        self.login_button.click()

    def ensure_authenticated(self, name: str, email: str, password: str):
        """
        Attempts login; if account is not registered or credentials fail,
        registers a new unique account via signup and logs in automatically.
        """
        self.goto()
        self.login(email, password)

        try:
            expect(self.page).to_have_url(re.compile(r".*/dashboard"), timeout=8000)
            return  # Login succeeded – done
        except Exception:
            pass  # Login failed – proceed to signup

        # Generate a unique email using timestamp to avoid "already registered" conflicts
        timestamp = int(time.time())
        unique_email = f"pnexgt22005+qa{timestamp}@gmail.com"

        self.navigate_to("/signup")
        expect(self.page.get_by_role("heading", name=re.compile(r"Create your WebMind account", re.IGNORECASE))).to_be_visible(timeout=10000)

        self.page.get_by_test_id("signup-name").fill(name)
        self.page.get_by_test_id("signup-email").fill(unique_email)
        self.page.get_by_test_id("signup-password").fill(password)
        # Wait briefly for password-strength checklist to settle before filling confirm
        self.page.wait_for_timeout(500)
        self.page.get_by_test_id("signup-confirm-password").fill(password)
        self.page.wait_for_timeout(500)
        self.page.get_by_test_id("signup-submit-button").click()

        expect(self.page).to_have_url(re.compile(r".*/dashboard"), timeout=15000)

    def assert_login_success(self):
        """Assert user is redirected to Dashboard upon successful login."""
        expect(self.page).to_have_url(re.compile(r".*/dashboard"), timeout=15000)

    def assert_error_message(self, expected_message: str):
        """Assert error message alert is displayed with expected text."""
        error_element = self.page.get_by_text(expected_message)
        expect(error_element).to_be_visible(timeout=8000)

File: QA-Assignment/Automation/pages/LoginPage.js
import { expect } from '@playwright/test';

/**
 * Page Object Model for the Login Page (/login)
 */
export class LoginPage {
  /**
   * @param {import('@playwright/test').Page} page
   */
  constructor(page) {
    this.page = page;

    // Accessible and reliable locators matching existing WebMind UI
    this.heading = page.getByRole('heading', { name: 'Account Login' });
    this.emailInput = page.getByTestId('login-email');
    this.passwordInput = page.getByTestId('login-password');
    this.loginButton = page.getByTestId('login-button');
    this.signupButton = page.getByTestId('signup-button');
  }

  /**
   * Navigate to Login page
   */
  async goto() {
    await this.page.goto('/login');
    await expect(this.heading).toBeVisible({ timeout: 15000 });
  }

  /**
   * Perform login action
   * @param {string} email
   * @param {string} password
   */
  async login(email, password) {
    await this.emailInput.clear();
    if (email !== undefined && email !== null && email !== '') {
      await this.emailInput.fill(email);
    }
    await this.passwordInput.clear();
    if (password !== undefined && password !== null && password !== '') {
      await this.passwordInput.fill(password);
    }
    await this.loginButton.click();
  }

  /**
   * Ensure authenticated: try login first, fallback to signup with unique email.
   * This handles the case where the configured credentials don't exist on the server.
   * @param {string} name
   * @param {string} email
   * @param {string} password
   */
  async ensureAuthenticated(name, email, password) {
    await this.goto();
    await this.login(email, password);

    try {
      await expect(this.page).toHaveURL(/.*\/dashboard/, { timeout: 8000 });
      return; // Login succeeded
    } catch {
      // Login failed – fallback to signup with unique timestamped email
    }

    const timestamp = Date.now();
    const uniqueEmail = `pnexgt22005+qa${timestamp}@gmail.com`;

    await this.page.goto('/signup');
    await expect(this.page.getByRole('heading', { name: /Create your WebMind account/i })).toBeVisible({ timeout: 10000 });

    await this.page.getByTestId('signup-name').fill(name);
    await this.page.getByTestId('signup-email').fill(uniqueEmail);
    await this.page.getByTestId('signup-password').fill(password);
    // Brief pause for password-strength UI to settle before filling confirm
    await this.page.waitForTimeout(500);
    await this.page.getByTestId('signup-confirm-password').fill(password);
    await this.page.waitForTimeout(500);
    await this.page.getByTestId('signup-submit-button').click();

    await expect(this.page).toHaveURL(/.*\/dashboard/, { timeout: 15000 });
  }

  /**
   * Assert login is successful and redirects to dashboard
   */
  async assertLoginSuccess() {
    await expect(this.page).toHaveURL(/.*\/dashboard/, { timeout: 15000 });
  }

  /**
   * Assert error message is displayed on screen
   * @param {string} expectedMessage
   */
  async assertErrorMessage(expectedMessage) {
    const errorElement = this.page.getByText(expectedMessage);
    await expect(errorElement).toBeVisible({ timeout: 8000 });
  }

  /**
   * Clear input fields
   */
  async clearFields() {
    await this.emailInput.clear();
    await this.passwordInput.clear();
  }
}
